### ISSUE TITLE
`pod repo push` no longer works after `repo-art add`

### ISSUE THREAD

**Issue by Ashton-W** [jfrog/cocoapods-art#6]
It seems the existence of artifactory repos on the system causes an error in the `pod repo push` commands.

```
$ bundle exec pod repo push REDACTEDSpecs REDACTED.podspec --private

Validating spec
Cloning spec repo `-1` from ``
[!] Unable to add a source with url `` named `-1`.
You can try adding it manually in `~/.cocoapods/repos` or via `pod repo add`.
```

<details>
<summary>

Click to toggle verbose output</summary>

```
$ bundle exec pod repo push REDACTEDSpecs REDACTED.podspec --private --verbose

Validating spec

Cloning spec repo `-1` from ``
  $ /usr/local/bin/git clone  -1
  error: unknown switch `1'
  usage: git clone [<options>] [--] <repo> [<dir>]

      -v, --verbose         be more verbose
      -q, --quiet           be more quiet
      --progress            force progress reporting
      -n, --no-checkout     don't create a checkout
      --bare                create a bare repository
      --mirror              create a mirror repository (implies bare)
      -l, --local           to clone from a local repository
      --no-hardlinks        don't use local hardlinks, always copy
      -s, --shared          setup as shared repository
      --recursive           initialize submodules in the clone
      --recurse-submodules  initialize submodules in the clone
      -j, --jobs <n>        number of submodules cloned in parallel
      --template <template-directory>
                            directory from which templates will be used
      --reference <repo>    reference repository
      --dissociate          use --reference only while cloning
      -o, --origin <name>   use <name> instead of 'origin' to track upstream
      -b, --branch <branch>
                            checkout <branch> instead of the remote's HEAD
      -u, --upload-pack <path>
                            path to git-upload-pack on the remote
      --depth <depth>       create a shallow clone of that depth
      --single-branch       clone only one branch, HEAD or --branch
      --shallow-submodules  any cloned submodules will be shallow
      --separate-git-dir <gitdir>
                            separate git dir from working tree
      -c, --config <key=value>
                            set config inside the new repository
      -4, --ipv4            use IPv4 addresses only
      -6, --ipv6            use IPv6 addresses only

[!] Unable to add a source with url `` named `-1`.
(/usr/local/bin/git clone  -1

error: unknown switch `1'
usage: git clone [<options>] [--] <repo> [<dir>]

    -v, --verbose         be more verbose
    -q, --quiet           be more quiet
    --progress            force progress reporting
    -n, --no-checkout     don't create a checkout
    --bare                create a bare repository
    --mirror              create a mirror repository (implies bare)
    -l, --local           to clone from a local repository
    --no-hardlinks        don't use local hardlinks, always copy
    -s, --shared          setup as shared repository
    --recursive           initialize submodules in the clone
    --recurse-submodules  initialize submodules in the clone
    -j, --jobs <n>        number of submodules cloned in parallel
    --template <template-directory>
                          directory from which templates will be used
    --reference <repo>    reference repository
    --dissociate          use --reference only while cloning
    -o, --origin <name>   use <name> instead of 'origin' to track upstream
    -b, --branch <branch>
                          checkout <branch> instead of the remote's HEAD
    -u, --upload-pack <path>
                          path to git-upload-pack on the remote
    --depth <depth>       create a shallow clone of that depth
    --single-branch       clone only one branch, HEAD or --branch
    --shallow-submodules  any cloned submodules will be shallow
    --separate-git-dir <gitdir>
                          separate git dir from working tree
    -c, --config <key=value>
                          set config inside the new repository
    -4, --ipv4            use IPv4 addresses only
    -6, --ipv6            use IPv6 addresses only

)
You can try adding it manually in `~/.cocoapods/repos` or via `pod repo add`.

/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/sources_manager.rb:35:in `rescue in find_or_create_source_with_url'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/sources_manager.rb:38:in `find_or_create_source_with_url'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/sources_manager.rb:58:in `source_with_name_or_url'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/validator.rb:33:in `block in initialize'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/validator.rb:33:in `map'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/validator.rb:33:in `initialize'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/command/repo/push.rb:120:in `new'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/command/repo/push.rb:120:in `block in validate_podspec_files'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/command/repo/push.rb:119:in `each'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/command/repo/push.rb:119:in `validate_podspec_files'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/command/repo/push.rb:64:in `run'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/claide-1.0.0/lib/claide/command.rb:334:in `run'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/lib/cocoapods/command.rb:50:in `run'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/lib/ruby/gems/2.1.0/gems/cocoapods-1.0.1/bin/pod:55:in `<top (required)>'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/bin/pod:23:in `load'
/Users/AshtonWilliams/.rbenv/versions/2.1.3/bin/pod:23:in `<main>'
```

</details>

---

<details>
<summary>

Click to toggle contents of `Gemfile.lock`</summary>

```
GEM
  remote: REDACTED
  specs:
    activesupport (4.2.6)
      i18n (~> 0.7)
      json (~> 1.7, >= 1.7.7)
      minitest (~> 5.1)
      thread_safe (~> 0.3, >= 0.3.4)
      tzinfo (~> 1.1)
    addressable (2.4.0)
    babosa (1.0.2)
    cert (1.4.1)
      fastlane_core (>= 0.29.1, < 1.0.0)
      spaceship (>= 0.22.0, < 1.0.0)
    claide (1.0.0)
    clamp (0.6.5)
    cocoapods (1.0.1)
      activesupport (>= 4.0.2)
      claide (>= 1.0.0, < 2.0)
      cocoapods-core (= 1.0.1)
      cocoapods-deintegrate (>= 1.0.0, < 2.0)
      cocoapods-downloader (>= 1.0.0, < 2.0)
      cocoapods-plugins (>= 1.0.0, < 2.0)
      cocoapods-search (>= 1.0.0, < 2.0)
      cocoapods-stats (>= 1.0.0, < 2.0)
      cocoapods-trunk (>= 1.0.0, < 2.0)
      cocoapods-try (>= 1.0.0, < 2.0)
      colored (~> 1.2)
      escape (~> 0.0.4)
      fourflusher (~> 0.3.0)
      molinillo (~> 0.4.5)
      nap (~> 1.0)
      xcodeproj (>= 1.1.0, < 2.0)
    cocoapods-art (0.9.5.pre)
    cocoapods-core (1.0.1)
      activesupport (>= 4.0.2)
      fuzzy_match (~> 2.0.4)
      nap (~> 1.0)
    cocoapods-deintegrate (1.0.0)
    cocoapods-downloader (1.0.1)
    cocoapods-plugins (1.0.0)
      nap
    cocoapods-search (1.0.0)
    cocoapods-stats (1.0.0)
    cocoapods-trunk (1.0.0)
      nap (>= 0.8, < 2.0)
      netrc (= 0.7.8)
    cocoapods-try (1.0.0)
    cocoapods-update-if-you-dare (0.1.1)
    colored (1.2)
    commander (4.3.5)
      highline (~> 1.7.2)
    credentials_manager (0.16.0)
      colored
      commander (>= 4.3.5)
      highline (>= 1.7.1)
      security
    deliver (1.13.1)
      credentials_manager (>= 0.16.0, < 1.0.0)
      fastimage (~> 1.6)
      fastlane_core (>= 0.46.2, < 1.0.0)
      plist (~> 3.1.0)
      spaceship (>= 0.26.2, < 1.0.0)
    domain_name (0.5.20160615)
      unf (>= 0.0.5, < 1.0.0)
    dotenv (2.1.1)
    escape (0.0.4)
    excon (0.45.4)
    faraday (0.9.2)
      multipart-post (>= 1.2, < 3)
    faraday-cookie_jar (0.0.6)
      faraday (>= 0.7.4)
      http-cookie (~> 1.0.0)
    faraday_middleware (0.10.0)
      faraday (>= 0.7.4, < 0.10)
    fastimage (1.6.8)
      addressable (~> 2.3, >= 2.3.5)
    fastlane (1.89.0)
      addressable (~> 2.3)
      cert (>= 1.4.1, < 2.0.0)
      credentials_manager (>= 0.16.0, < 1.0.0)
      deliver (>= 1.11.2, < 2.0.0)
      fastlane_core (>= 0.43.3, < 1.0.0)
      frameit (>= 2.6.2, < 3.0.0)
      gym (>= 1.6.3, < 2.0.0)
      krausefx-shenzhen (>= 0.14.7)
      match (>= 0.5.0, < 1.0.0)
      multipart-post (~> 2.0.0)
      pem (>= 1.3.1, < 2.0.0)
      pilot (>= 1.5.1, < 2.0.0)
      plist (~> 3.1.0)
      produce (>= 1.1.2, < 2.0.0)
      scan (>= 0.6.1, < 1.0.0)
      screengrab (>= 0.3.2, < 1.0.0)
      sigh (>= 1.8.0, < 2.0.0)
      slack-notifier (~> 1.3)
      snapshot (>= 1.12.1, < 2.0.0)
      spaceship (>= 0.26.2, < 1.0.0)
      supply (>= 0.6.2, < 1.0.0)
      terminal-notifier (~> 1.6.2)
      terminal-table (~> 1.4.5)
      xcodeproj (>= 0.20, < 2.0.0)
      xcpretty (>= 0.2.1)
    fastlane_core (0.48.0)
      babosa
      colored
      commander (= 4.3.5)
      credentials_manager (>= 0.16.0, < 1.0.0)
      excon (~> 0.45.0)
      gh_inspector (>= 1.0.1, < 2.0.0)
      highline (>= 1.7.2)
      json
      multi_json
      plist (~> 3.1)
      rubyzip (~> 1.1.6)
      terminal-table (~> 1.4.5)
    fourflusher (0.3.2)
    frameit (2.7.0)
      deliver (> 0.3)
      fastimage (~> 1.6.3)
      fastlane_core (>= 0.36.1, < 1.0.0)
      mini_magick (~> 4.5.1)
    fuzzy_match (2.0.4)
    gh_inspector (1.0.2)
    google-api-client (0.9.9)
      addressable (~> 2.3)
      googleauth (~> 0.5)
      httpclient (~> 2.7)
      hurley (~> 0.1)
      memoist (~> 0.11)
      mime-types (>= 1.6)
      representable (~> 2.3.0)
      retriable (~> 2.0)
      thor (~> 0.19)
    googleauth (0.5.1)
      faraday (~> 0.9)
      jwt (~> 1.4)
      logging (~> 2.0)
      memoist (~> 0.12)
      multi_json (~> 1.11)
      os (~> 0.9)
      signet (~> 0.7)
    gym (1.6.3)
      fastlane_core (>= 0.43.1, < 1.0.0)
      plist
      rubyzip (>= 1.1.7)
      terminal-table
      xcpretty (>= 0.2.1)
    highline (1.7.8)
    http-cookie (1.0.2)
      domain_name (~> 0.5)
    httpclient (2.8.0)
    hurley (0.2)
    i18n (0.7.0)
    json (1.8.3)
    jwt (1.5.4)
    krausefx-shenzhen (0.14.9)
      commander (~> 4.3)
      dotenv (>= 0.7)
      faraday (~> 0.9)
      faraday_middleware (~> 0.9)
      highline (>= 1.7.2)
      json (~> 1.8)
      net-sftp (~> 2.1.2)
      plist (~> 3.1.0)
      rubyzip (~> 1.1)
      security (~> 0.1.3)
      terminal-table (~> 1.4.5)
    little-plugger (1.1.4)
    logging (2.1.0)
      little-plugger (~> 1.1)
      multi_json (~> 1.10)
    match (0.6.1)
      cert (>= 1.2.8, < 2.0.0)
      credentials_manager (>= 0.13.0, < 1.0.0)
      fastlane_core (>= 0.39.0, < 1.0.0)
      security
      sigh (>= 1.2.2, < 2.0.0)
      spaceship (>= 0.24.0, < 1.0.0)
    memoist (0.14.0)
    mime-types (3.1)
      mime-types-data (~> 3.2015)
    mime-types-data (3.2016.0521)
    mini_magick (4.5.1)
    mini_portile2 (2.1.0)
    minitest (5.9.0)
    molinillo (0.4.5)
    multi_json (1.12.1)
    multi_xml (0.5.5)
    multipart-post (2.0.0)
    nap (1.1.0)
    net-sftp (2.1.2)
      net-ssh (>= 2.6.5)
    net-ssh (3.2.0)
    netrc (0.7.8)
    nokogiri (1.6.8)
      mini_portile2 (~> 2.1.0)
      pkg-config (~> 1.1.7)
    os (0.9.6)
    pem (1.3.1)
      fastlane_core (>= 0.43.1, < 1.0.0)
      spaceship (>= 0.26.2, < 1.0.0)
    pilot (1.9.2)
      credentials_manager (>= 0.3.0)
      fastlane_core (>= 0.46.2, < 1.0.0)
      spaceship (>= 0.27.0, < 1.0.0)
      terminal-table (~> 1.4.5)
    pkg-config (1.1.7)
    plist (3.1.0)
    produce (1.1.2)
      fastlane_core (>= 0.30.0, < 1.0.0)
      spaceship (>= 0.16.0, < 1.0.0)
    representable (2.3.0)
      uber (~> 0.0.7)
    retriable (2.1.0)
    rouge (1.11.1)
    rubyzip (1.1.7)
    scan (0.8.0)
      fastlane_core (>= 0.36.1, < 1.0.0)
      slack-notifier (~> 1.3)
      terminal-table
      xcpretty (>= 0.2.1)
      xcpretty-travis-formatter (>= 0.0.3)
    screengrab (0.3.2)
      fastlane_core (>= 0.38.0, < 1.0.0)
    security (0.1.3)
    sigh (1.8.0)
      fastlane_core (>= 0.36.1, < 1.0.0)
      plist (~> 3.1)
      spaceship (>= 0.22.0, < 1.0.0)
    signet (0.7.3)
      addressable (~> 2.3)
      faraday (~> 0.9)
      jwt (~> 1.5)
      multi_json (~> 1.10)
    slack-notifier (1.5.1)
    slather (2.2.1)
      clamp (~> 0.6)
      nokogiri (~> 1.6.3)
      xcodeproj (~> 1.1)
    snapshot (1.12.3)
      fastimage (~> 1.6.3)
      fastlane_core (>= 0.36.1, < 1.0.0)
      plist (~> 3.1.0)
      xcpretty (>= 0.2.1)
    spaceship (0.27.2)
      colored
      credentials_manager (>= 0.9.0)
      faraday (~> 0.9)
      faraday-cookie_jar (~> 0.0.6)
      faraday_middleware (~> 0.9)
      fastimage (~> 1.6)
      multi_xml (~> 0.5)
      plist (~> 3.1)
    supply (0.7.0)
      credentials_manager (>= 0.15.0)
      fastlane_core (>= 0.43.4)
      google-api-client (~> 0.9.1)
    terminal-notifier (1.6.3)
    terminal-table (1.4.5)
    thor (0.19.1)
    thread_safe (0.3.5)
    tzinfo (1.2.2)
      thread_safe (~> 0.1)
    uber (0.0.15)
    unf (0.1.4)
      unf_ext
    unf_ext (0.0.7.2)
    xcodeproj (1.1.0)
      activesupport (>= 3)
      claide (>= 1.0.0, < 2.0)
      colored (~> 1.2)
    xcpretty (0.2.2)
      rouge (~> 1.8)
    xcpretty-travis-formatter (0.0.4)
      xcpretty (~> 0.2, >= 0.0.7)

PLATFORMS
  ruby

DEPENDENCIES
  cocoapods (~> 1.0.1)
  cocoapods-art (~> 0.9.5.pre)
  cocoapods-update-if-you-dare (~> 0.1.1)
  fastlane (~> 1.89.0)
  slather (~> 2.2.1)
  xcpretty (~> 0.2.2)

BUNDLED WITH
   1.11.2
```

</details>

**Comment by 85danf** [comment]
Hi,
Can you please post the following:
1. contents of your `Podfile`
2. output of `pod repo`
3. output of `pod repo-art`
4. output of `gem list | grep coco`

**Comment by Ashton-W** [comment]
<details>
<summary>

Click to toggle `Podfile`</summary>



``` ruby
ENV["COCOAPODS_DISABLE_STATS"] = "1"

plugin 'cocoapods-art', :sources => [
  'REDACTED-public', 'REDACTED-internal'
]

platform :ios, '8.0'

project 'REDACTED'

use_frameworks!

target 'REDACTED' do
  podspec
end

target 'UnitTests' do
  podspec

  # Third Party Dependencies
  pod 'Specta', '1.0.2', :inhibit_warnings => true
  pod 'Expecta', '1.0.0', :inhibit_warnings => true
  pod 'OCMockito', '2.0.1', :inhibit_warnings => true
  pod 'Nocilla', :inhibit_warnings => true
end

target 'IntegrationTests' do
  podspec

  # Third Party Dependencies
  pod 'Specta', '1.0.2', :inhibit_warnings => true
  pod 'Expecta', '1.0.0', :inhibit_warnings => true
  pod 'OCMockito', '2.0.1', :inhibit_warnings => true
  pod 'Nocilla', :inhibit_warnings => true
end
```

</details>

<details>
<summary>

Click to toggle `pod repo`</summary>



``` yaml

REDACTED-public
- Type: local
- URL:  
- Path: /Users/AshtonWilliams/.cocoapods/repos/REDACTED-public

master
- Type: git (master)
- URL:  https://github.com/CocoaPods/Specs.git
- Path: /Users/AshtonWilliams/.cocoapods/repos/master

REDACTED-internal
- Type: local
- URL:  
- Path: /Users/AshtonWilliams/.cocoapods/repos/REDACTED-internal

REDACTED-REDACTEDSpecs
- Type: git (master)
- URL:  https://github.com/CocoaPods/Specs.git
- Path: /Users/AshtonWilliams/.cocoapods/repos/REDACTED-REDACTEDSpecs

4 repos
```

</details>

<details>
<summary>

Click to toggle `pod repo-art`</summary>



``` yaml

REDACTED-public
- Type: Artifactory
- URL:  http://artifactory.REDACTED.com/artifactory/api/pods/REDACTED-public
- Path: /Users/AshtonWilliams/.cocoapods/repos/REDACTED-public

REDACTED-internal
- Type: Artifactory
- URL:  http://artifactory.REDACTED.com/artifactory/api/pods/REDACTED-internal
- Path: /Users/AshtonWilliams/.cocoapods/repos/REDACTED-internal

2 repos
```

</details>

<details>
<summary>

Click to toggle `gem list | grep coco`</summary>



```
cocoapods (1.0.1, 1.0.0, 1.0.0.rc.2, 1.0.0.beta.8, 0.39.0, 0.37.2, 0.37.1)
cocoapods-art (0.9.5.pre)
cocoapods-core (1.0.1, 1.0.0, 1.0.0.rc.2, 1.0.0.beta.8, 0.39.0, 0.37.2, 0.37.1)
cocoapods-deintegrate (1.0.0, 1.0.0.rc.2, 1.0.0.beta.1)
cocoapods-downloader (1.0.1, 1.0.0, 1.0.0.rc.1, 1.0.0.beta.3, 0.9.3, 0.9.0)
cocoapods-keys (1.6.1, 1.6.0)
cocoapods-plugins (1.0.0, 1.0.0.rc.1, 1.0.0.beta.1, 0.4.2)
cocoapods-search (1.0.0, 1.0.0.rc.1, 1.0.0.beta.2, 0.1.0)
cocoapods-stats (1.0.0, 1.0.0.rc.1, 1.0.0.beta.4, 0.6.2)
cocoapods-trunk (1.0.0, 1.0.0.rc.1, 1.0.0.beta.4, 0.6.4, 0.6.0)
cocoapods-try (1.0.0, 1.0.0.rc.1, 1.0.0.beta.4, 0.5.1, 0.4.5, 0.4.4)
cocoapods-update-if-you-dare (0.1.1, 0.1.0)
```

Note: We are using bundler, `Gemfile.lock` was posted earlier.
Note: `cocoapods-art (0.9.5.pre)` is a build of this repos `develop` branch.
</details>

Excuse the redactions.

**Comment by svenresch** [comment]
Same. After executing a 'pod repo-art add' command the resulting repo is listed with an empty URL property and this seems to be the issue.

**Comment by 85danf** [comment]
Actually calling `pod repo` will always output repos with empty urls - this is due to the fact that Cocoapods looks for the .git/config file in repos for the url (which cocoapods-art does not use).
When you run `pod repo-art` however cocoapods-art searches for the url in the special file we put there (called `.artpodrc` under the each repo's root) - this is the same url cocoapods-art uses to resolve its remote.

I did however notice that you have 2 master repos (which point to `https://github.com/CocoaPods/Specs.git`) -- during our tests we noticed the linter might have issues with such scenarios, although you named them differently locally.

Last but not least, notice the many versions of Cocoapods you have installed - this is an inconsistent state which will cause you issues for sure (meaning which version of each gem runs is undefined) - we've already seen users encounter inconsistent behavior in such as state (see issue #2 for example).

I suggest you:
1. `gem uninstall` all cocoapods gems and cocoapods-art as well.
2. `gem install` cocoapods
3. `gem install` cocoapods-art
4. remove the extra master specs repo (either master or REDACTED-REDACTEDSpecs)

I believe it should solve your issue.

**Comment by 85danf** [comment]
Closed due to inactivity

**Comment by mirkoluchi** [comment]
Hi.
The issue is still present.
It seems to me to be related to the way `pod repo push` works, by fetching the specs repos listed  to perform the linting of the podspec being pushed.
Whenever it tries to retrieve the Artifactory repo, it fails, because it has no URL (since it is not a Git repo).
Please help.

The only workaround I found right now is to manually push the podspec to the private repo, using just Git push, instead of pod repo push (which however doesn't allow me to validate the podspec being pushed).

**Comment by mirkoluchi** [comment]
I found this workaround: creating a local git repository in the directory where the Artifactory repo is placed solves the issue.

Before the workaround, my Artifactory repository is listed by `pod repo list` like this:

```
mirko_art_repo
- Type: local
- URL:  
- Path: /Users/mluchi/.cocoapods/repos/mirko_art_repo
```

My workaround is applied in this way:

```
cd ~/.cocoapods/repos/mirko_art_repo
git init
```

After the workaround, `pod repo list` returns this:

```
mirko_art_repo
- Type: git (unknown)
- URL:  file:///Users/mluchi/.cocoapods/repos/mirko_art_repo/.git
- Path: /Users/mluchi/.cocoapods/repos/mirko_art_repo
```

And therefore, since the URL is no longer empty, `pod repo push` is capable of performing the linting with no error.

**Comment by 85danf** [comment]
Thanks for the info @mirkoluchi, We will check this and get back to you

**Comment by Ashton-W** [comment]
@feldan would it be acceptable to have the plugin run `git init` in the artifactory repo directories automatically? This workaround from @mirkoluchi has solved this problem with no adverse affects for us.

**Comment by mirkoluchi** [comment]
Hi, even if the empty Git repo fixes the issue when performing `pod repo push`, it introduces issues when running `pod update`, with this error:

`
fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
`

The only workaround I found to make `pod update` work is to manually update the pod repo(s) and then run `pod update --no-repo-update`.

@jfrog Could you please find a complete fix to the issues with your Cocoapods plugin?

**Comment by 85danf** [comment]
@Ashton-W @mirkoluchi 
The local repos the plugin creates are not meant to serve as git repos, for that reason `pod repo` will not show the git url, while `pod repo-art` will show the remote urls as it is implementation we have in the plugin only. 
So, to answer @svenresch 's question as well, here is a sample output on my machine:

```
art-local
- Type: Artifactory
- URL:  http://localhost:8080/artifactory/api/pods/coco-local
- Path: /Users/me/.cocoapods/repos/art-local```

```
art-local
- Type: local
- URL:
- Path: /Users/me/.cocoapods/repos/art-local```

This is because cocoapods look for the git info under .git, while repo-art looks for url info stored under the `.artpodspec` file it stores in the local repo.

Running `git init` under an artifactory repo will simply cause Cocoapods to think it's a git repo and of course fail (as there's no [remote] info in that git repo).

**Comment by mirkoluchi** [comment]
@feldan 

Hi. Yes, that was clear. In fact running `git init` within the repo folder was a workaround to make the Artifactory local repo look like a git one, which was necessary for making the `pod repo push` command work.
But since it introduces issues when running `pod update`, it's not good.

So, forgetting about the workaround and going back to the initial question of this thread, **could you please tell us how to make the `pod repo push `command work, when any Artifactory repo is present on the machine?**

Just to recap the issue:
- I have both a private Git-based podspec repo (for source distribution of my pods) and an Artifactory repo (for binary distribution of my pods)

`pod repo list`

```
myartifactory-repo
fatal: Not a git repository (or any of the parent directories): .git
- Type: local
- URL:  
- Path: /Users/myuser/.cocoapods/repos/myartifactory-repo

master
- Type: git (master)
- URL:  https://github.com/CocoaPods/Specs.git
- Path: /Users/myuser/.cocoapods/repos/master

mygit-repo
- Type: git (master)
- URL:  git@bitbucket.org:mygitrepo.git
- Path: /Users/myuser/.cocoapods/repos/mygit-repo

3 repos
```


- Whenever I try to push a podspec to the Git-based podspec repo using the `pod repo push` command, it fails with this error:

`pod repo push --verbose --allow-warnings mygit-repo Framework.podspec`

```

Validating spec

Cloning spec repo `-1` from ``
  $ /usr/bin/git clone  -1
  error: unknown switch `1'
  usage: git clone [<options>] [--] <repo> [<dir>]
  
      -v, --verbose         be more verbose
      -q, --quiet           be more quiet
      --progress            force progress reporting
      -n, --no-checkout     don't create a checkout
      --bare                create a bare repository
      --mirror              create a mirror repository (implies bare)
      -l, --local           to clone from a local repository
      --no-hardlinks        don't use local hardlinks, always copy
      -s, --shared          setup as shared repository
      --recursive           initialize submodules in the clone
      --recurse-submodules  initialize submodules in the clone
      -j, --jobs <n>        number of submodules cloned in parallel
      --template <template-directory>
                            directory from which templates will be used
      --reference <repo>    reference repository
      --dissociate          use --reference only while cloning
      -o, --origin <name>   use <name> instead of 'origin' to track upstream
      -b, --branch <branch>
                            checkout <branch> instead of the remote's HEAD
      -u, --upload-pack <path>
                            path to git-upload-pack on the remote
      --depth <depth>       create a shallow clone of that depth
      --single-branch       clone only one branch, HEAD or --branch
      --shallow-submodules  any cloned submodules will be shallow
      --separate-git-dir <gitdir>
                            separate git dir from working tree
      -c, --config <key=value>
                            set config inside the new repository
      -4, --ipv4            use IPv4 addresses only
      -6, --ipv6            use IPv6 addresses only
  
[!] Unable to add a source with url `` named `-1`.
(/usr/bin/git clone  -1

error: unknown switch `1'
usage: git clone [<options>] [--] <repo> [<dir>]

    -v, --verbose         be more verbose
    -q, --quiet           be more quiet
    --progress            force progress reporting
    -n, --no-checkout     don't create a checkout
    --bare                create a bare repository
    --mirror              create a mirror repository (implies bare)
    -l, --local           to clone from a local repository
    --no-hardlinks        don't use local hardlinks, always copy
    -s, --shared          setup as shared repository
    --recursive           initialize submodules in the clone
    --recurse-submodules  initialize submodules in the clone
    -j, --jobs <n>        number of submodules cloned in parallel
    --template <template-directory>
                          directory from which templates will be used
    --reference <repo>    reference repository
    --dissociate          use --reference only while cloning
    -o, --origin <name>   use <name> instead of 'origin' to track upstream
    -b, --branch <branch>
                          checkout <branch> instead of the remote's HEAD
    -u, --upload-pack <path>
                          path to git-upload-pack on the remote
    --depth <depth>       create a shallow clone of that depth
    --single-branch       clone only one branch, HEAD or --branch
    --shallow-submodules  any cloned submodules will be shallow
    --separate-git-dir <gitdir>
                          separate git dir from working tree
    -c, --config <key=value>
                          set config inside the new repository
    -4, --ipv4            use IPv4 addresses only
    -6, --ipv6            use IPv6 addresses only

)
You can try adding it manually in `~/.cocoapods/repos` or via `pod repo add`.

/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/sources_manager.rb:35:in `rescue in find_or_create_source_with_url'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/sources_manager.rb:38:in `find_or_create_source_with_url'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/sources_manager.rb:58:in `source_with_name_or_url'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/validator.rb:31:in `block in initialize'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/validator.rb:31:in `map'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/validator.rb:31:in `initialize'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:122:in `new'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:122:in `block in validate_podspec_files'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:121:in `each'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:121:in `validate_podspec_files'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:66:in `run'
/Library/Ruby/Gems/2.0.0/gems/claide-1.0.1/lib/claide/command.rb:334:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command.rb:50:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/bin/pod:55:in `<top (required)>'
/usr/local/bin/pod:23:in `load'
/usr/local/bin/pod:23:in `<main>'
```

The issue is caused by the podspec validation made by the command, which tries to update any podspec repo found (included the Artifactory one), but - since the Artifactory repo has no Git URL - fails.

**Comment by eyalbe4** [comment]
@mirkoluchi,
We are working on finding a solution for this issue. We'll send out an update soon.

**Comment by eyalbe4** [comment]
All, just updating that we came with a solution that should fix this issue.
It requires us to implement a few changes in the plugin implementation. The main change is moving the Artifactory repositories out of the *~/.cocoapods/repos/* directory into a new directory.
We'll keep you guys posted.

**Comment by vexorsis** [comment]
have the same issue

**Comment by eyalbe4** [comment]
@Ashton-W, @wangpeiyan and @mirkoluchi,
We pushed [this](https://github.com/JFrogDev/cocoapods-art/commit/482cda09365e9ff74a0246a5bd71a9f55ec9bf83) commit to fix this issue. The fix includes the following change:
Artifactory spec repos are no longer saved in *~/.cocoapods/repos*, but in a new directory: *~/.cocoapods/repos-art*. The *pod repo-art* commands syntax and functionality should not be effected by this change. 
After installing the new *cocoapods-art* plugin, you'll need to either move the Artifactory spec repos to the new directory, or add them using the *pod repo-art add* command (and deleting them from  *~/.cocoapods/repos*).
Before releasing the new gem, will you guys be able to verify that it indeed resolves the issue?
You can do it by 
1. Cloning this git repository.
2. Building the new cloned code: *gem build cocoapods_art.gemspec*
3. Uninstalling the old gem: *gem uninstall cocoapods-art*
4. Installing the new gem built: *gem install cocoapods-art-0.9.5.gem*

I appreciate your help with this and I'm looking forward for you feedback.

**Comment by Ashton-W** [comment]
Could we move the repo-art directories to the new location automatically?

**Comment by eyalbe4** [comment]
I guess we could @Ashton-W, but why would we need that? All we need to do is run one mv command for each spec repo (a one time migration).

**Comment by Ashton-W** [comment]
it's a one time migration for each member of my team and each CI server. Updating the gem will be a simple change in our `Gemfile.lock`, moving the specs will be manual on 30+ Macs

**Comment by eyalbe4** [comment]
Got it @Ashton-W. No problem. We'll add a shell script that will do the migration. I'll let you know when it is ready.

**Comment by mirkoluchi** [comment]
Hi,

I tested the fix, but it still have issues.
The issue now is that the `pod repo push` command is no longer able to validate the podspec being pushed, probably because it does not take into accounts Artifactory repos to find specs of depending pods.

In my test, I have a **Framework.podspec** which **depends on a pod named 'MyLibrary'**, which is **published on my artifactory repository**.
This is the command line output:


`
pod repo push mygit-repo Framework.podspec 


Validating spec
    Framework (2.0.3) - Analyzing on iOS 8.0 platform.
  Preparing

Analyzing dependencies

Inspecting targets to integrate
  Using `ARCHS` setting to build architectures of target `Pods-App`: (``)

Fetching external sources
-> Fetching podspec for `Framework` from `/Users/myuser/project/Framework.podspec`

Resolving dependencies of 
 -> Framework (2.0.3)
    - WARN  | source: Git SSH URLs will NOT work for people behind firewalls configured to only allow HTTP, therefore HTTPS is preferred.
    - ERROR | [iOS] unknown: Encountered an unknown error (Unable to find a specification for `MyLibrary` depended upon by `Framework`
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/resolver.rb:442:in `handle_resolver_error'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/resolver.rb:65:in `rescue in resolve'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/resolver.rb:57:in `resolve'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/analyzer.rb:723:in `block in resolve_dependencies'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/user_interface.rb:64:in `section'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/analyzer.rb:721:in `resolve_dependencies'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/analyzer.rb:76:in `analyze'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:236:in `analyze'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:150:in `block in resolve_dependencies'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/user_interface.rb:64:in `section'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:149:in `resolve_dependencies'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/validator.rb:403:in `block in download_pod'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/validator.rb:403:in `each'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/validator.rb:403:in `download_pod'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/validator.rb:299:in `block in perform_extensive_analysis'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/validator.rb:293:in `each'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/validator.rb:293:in `perform_extensive_analysis'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/validator.rb:79:in `validate'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:127:in `block in validate_podspec_files'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:121:in `each'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:121:in `validate_podspec_files'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:66:in `run'
/Library/Ruby/Gems/2.0.0/gems/claide-1.0.1/lib/claide/command.rb:334:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command.rb:50:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/bin/pod:55:in `<top (required)>'
/usr/local/bin/pod:23:in `load'
/usr/local/bin/pod:23:in `<main>'
) during validation.

[!] The `Framework.podspec` specification does not validate.

/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:132:in `block in validate_podspec_files'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:121:in `each'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:121:in `validate_podspec_files'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/repo/push.rb:66:in `run'
/Library/Ruby/Gems/2.0.0/gems/claide-1.0.1/lib/claide/command.rb:334:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command.rb:50:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/bin/pod:55:in `<top (required)>'
/usr/local/bin/pod:23:in `load'
/usr/local/bin/pod:23:in `<main>'

`

**Comment by eyalbe4** [comment]
@mirkoluchi,
We'd like to learn more about your development lifecycle.
Are you publishing pods to a github repository, while their dependencies are in Artifactory? Does the dependencies eventually make it to github? If so, when? Why do you manage some of your pods in an Artifactory repository and some in a git repository?

**Comment by mirkoluchi** [comment]
@eyalbe4 
Hi, we have a private Git repo (on bitbucket) that we use for source distribution of our libraries (for internal use).
We instead use Artifactory for the binary distribution of our libraries.
Both the 'source' and 'binary' podspecs sometimes refer to podspecs hosted on Github.

**Comment by eyalbe4** [comment]
@mirkoluchi, we may need to extend the cocoapods-art capabilities to support the use case you described and your feedback could really help us achieve that. Can you please let us know if performing the following steps allow you to successfully perform the push that failed for you?
1. Create a new (additional) git repository.
2. Add this new repository locally using *pod repo add*.
3. Download all podspec.json files from your Artifactory repository, and push them to the git repository using *pod repo push <repo name> <spe-name>.podspec.json*.
4. Run the *pod repo push* command which failed before.

**Comment by mirkoluchi** [comment]
@eyalbe4, I tried your suggestion and it works. 

This is what I did:
* I created a new Git repo (that will be a kind of Git-based copy of Artifactory podspecs) and added it to my local cocoapods repos
`pod repo add artifactory-git-repo <new git repo url> `

* I pushed the 'MyLibrary.podspec' (located in .cocoapods/repo-art/myartifactory-repo/MyLibrary/1.0/MyLibrary.podspec) to the newly created Git podspec repo 
`pod repo push artifactory-git-repo ~/.cocoapods/repo-art/myartifactory-repo/MyLibrary/1.0/MyLibrary.podspec`
The MyLibrary.podspec has the *source* attribute which refers to the HTTP URL where the MyLibrary.tar.gz can be GET: **spec.source = { :http => "https://my-artifactory-server.com/artifactory/api/pods/myartifactory-repo/pod/pkg/MyLibrary/1.0", :type => 'tgz'}
  end**


* I pushed the 'Framework.podspec' (that depends on 'MyLibrary' pod) to our private podspec Git repo
`pod repo push mygit-repo Framework.podspec`

The command worked fine, since the validation of the podspec was able to find the 'MyLibrary.podspec'  in the local .cocoapods/repos/artifactory-git-repo.

**Are you going to make the pod repo-art command automatically create this local Git repo containing a copy of the podspecs files downloaded from the Artifactory repo? Please let me know**

**Comment by mirkoluchi** [comment]
I tested locally the following solution.

**Please advice if that can be ok.**

1) when we first do `pod repo-art add artifactory-repo-name [artifactory-repo-url]`, your plugin could:
* create a new Git repository in the .cocoapods/repo-art/artifactory-repo-name directory and commit there all the podspecs downloaded from Artifactory
`mkdir .cocoapods/repo-art/artifactory-repo-name`
`cd .cocoapods/repo-art/artifactory-repo-name`
`git init`
`...curl to download specs from Artifactory...`
`git add .`
`git commit -m "Artifactory repo init"`

* create a Git repo  .cocoapods/repos/artifactory-repo-name which has the previous repository as remote
`cd .cocoapods/repos`
`git clone  file:///Users/myuser/.cocoapods/repo-art/artifactory-repo-name`

2) Each time the user executes `pod repo-art update artifactory-repo-name`, you should commit the newly downloaded specs:
`...curl to download updated specs from Artifactory into .cocoapods/repo-art/artifactory-repo-name...`
`cd .cocoapods/repo-art/artifactory-repo-name`
`git add .`
`git commit -m "Updated Artifactory specs"`


In this way, both `pod update` and `pod repo push` commands actually see the .cocoapods/repos/artifactory-repo-name as a valid Git repo with a valid remote (since it is really a Git repo with a remote), and everything uses the 'default' Cocoapods mechanism.

**Comment by nfalco79** [comment]
We have the same issue (licensed)
@mirkoluchi thank for the workaround but that procedure is not usable if not automated in the repo-art command or similar.

What is the JFrog use case designed to support artifactory cocoapods repository to be used with cocoapods commands if the basic commands to update or publish artifacts does not work (especially in continuos delivery scenario)?

**Comment by eyalbe4** [comment]
@nfalco79, we want to have the process handled automatically by the cocoapods-art commands.
@mirkoluchi, your idea looks good. Thanks a lot for sharing it. We'll work on implementing this as part of the plugin. 
We'll then ask you guys to try it out and provide your feedback.

**Comment by eyalbe4** [comment]
@mirkoluchi, thanks again for coming up and testing this solution!
We've pushed [this](https://github.com/JFrogDev/cocoapods-art/commit/c5a52386ac14367238164728a8189f20578b3594) commit with your solution's implementation.
Please try it out and let us what you think. Since the gem has not been released yet, you can build it from the sources by:
* Cloning this git repository.
* Building the new cloned code: gem build cocoapods_art.gemspec
* Uninstalling the old gem: gem uninstall cocoapods-art
* Installing the new gem built: gem install cocoapods-art-0.9.5.gem

Looking forward to your feedback.

**Comment by mirkoluchi** [comment]
@eyalbe4 I tested your patch. 
The `pod repo-art add` and `pod repo-art update` commands work fine. 
However the **`pod install`** and **`pod update`** commands **don't work correctly**.

This is how I did my test:
* I published a **Framework** pod on **my-artifactory-repo**
* I created a Podfile declaring both the **Artifactory** source  and the Cocoapods master **Git** source:
```
plugin 'cocoapods-art', :sources => [
    'my-artifactory-repo'
]
source 'https://github.com/CocoaPods/Specs'

platform :ios, 9.0
workspace 'workspace'
use_frameworks!

abstract_target 'App-dependencies' do
    pod 'Framework'
    target 'App' do
        project 'App/App.xcodeproj'
    end
end
```
* I tried to execute both `pod install` and `pod update`, and they both **fail** with error: `[!] Unable to find a specification for 'Framework'`

Full error stack (of `pod update --verbose`):
```
pod update --verbose
Update all pods
  Preparing

Updating local specs repositories

Updating spec repo `master`

CocoaPods 1.2.0 is available.
To update use: `sudo gem install cocoapods`

For more information, see https://blog.cocoapods.org and the CHANGELOG for this version at https://github.com/CocoaPods/CocoaPods/releases/tag/1.2.0


Analyzing dependencies

Inspecting targets to integrate
  Using `ARCHS` setting to build architectures of target `Pods-App-dependencies-App`: (``)

Resolving dependencies of `Podfile`
[!] Unable to find a specification for `Framework`

/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/resolver.rb:442:in `handle_resolver_error'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/resolver.rb:65:in `rescue in resolve'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/resolver.rb:57:in `resolve'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/analyzer.rb:723:in `block in resolve_dependencies'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/user_interface.rb:64:in `section'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/analyzer.rb:721:in `resolve_dependencies'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer/analyzer.rb:76:in `analyze'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:236:in `analyze'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:150:in `block in resolve_dependencies'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/user_interface.rb:64:in `section'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:149:in `resolve_dependencies'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/installer.rb:110:in `install!'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command/update.rb:60:in `run'
/Library/Ruby/Gems/2.0.0/gems/claide-1.0.1/lib/claide/command.rb:334:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/lib/cocoapods/command.rb:50:in `run'
/Library/Ruby/Gems/2.0.0/gems/cocoapods-1.1.1/bin/pod:55:in `<top (required)>'
/usr/local/bin/pod:23:in `load'
/usr/local/bin/pod:23:in `<main>'
```




**Notice:**
**If I remove the Git source** from the **Podfile**, everything works fine, but of course this is not a suitable solution, since I typically have dependencies from several sources, including Git repos (Cocoapods master repo or private podspec repos).
```
plugin 'cocoapods-art', :sources => [
    'my-artifactory-repo'
]
# source 'https://github.com/CocoaPods/Specs' <<--- If I remove this it works

platform :ios, 9.0
workspace 'workspace'
use_frameworks!

abstract_target 'App-dependencies' do
    pod 'Framework'
    target 'App' do
        project 'App/App.xcodeproj'
    end
end
```
Notice that in this case the console output shows that the Artifactory repo is being processed (it did not happen in the above scenario):
```
pod update --verbose
Update all pods
  Preparing
/Users/mluchi/.cocoapods/repos-art/my-artifactory-repo/.artpodrc

Updating local specs repositories
/Users/mluchi/.cocoapods/repos-art/my-artifactory-repo/.artpodrc

Updating spec repo `my-artifactory-repo`

Analyzing dependencies

Inspecting targets to integrate
  Using `ARCHS` setting to build architectures of target `Pods-App-dependencies-App`: (``)

Resolving dependencies of `Podfile`
/Users/mluchi/.cocoapods/repos-art/my-artifactory-repo/.artpodrc

Comparing resolved specification to the sandbox manifest
  - Framework

Downloading dependencies

-> Using Framework (1.0.0)
  - Running pre install hooks

Generating Pods project
  - Creating Pods project
  - Adding source files to Pods project
  - Adding frameworks to Pods project
  - Adding libraries to Pods project
  - Adding resources to Pods project
  - Linking headers
  - Installing targets
    - Installing target `Pods-App-dependencies-App` iOS 9.0
      - Generating Info.plist file at `Pods/Target Support Files/Pods-App-dependencies-App/Info.plist`
      - Generating module map file at `Pods/Target Support Files/Pods-App-dependencies-App/Pods-App-dependencies-App.modulemap`
      - Generating umbrella header at `Pods/Target Support Files/Pods-App-dependencies-App/Pods-App-dependencies-App-umbrella.h`
  - Running post install hooks
  - Writing Xcode project file to `Pods/Pods.xcodeproj`
    - Generating deterministic UUIDs
  - Writing Lockfile in `Podfile.lock`
  - Writing Manifest in `Pods/Manifest.lock`

Integrating client project

Integrating target `Pods-App-dependencies-App` (`App/App.xcodeproj` project)
  - Running post install hooks
    - cocoapods-stats from `/Library/Ruby/Gems/2.0.0/gems/cocoapods-stats-1.0.0/lib/cocoapods_plugin.rb`

Sending stats

-> Pod installation complete! There is 1 dependency from the Podfile and 1 total pod installed.

```

**Comment by Ashton-W** [comment]
@mirkoluchi FWIW we use all repo-art sources, we have a repo-art that is a mirror of CocoaPods/Specs. Don't know if your setup is valid but hope this helps

**Comment by mirkoluchi** [comment]
@Ashton-W Hi, we have to be able to mix repo-art sources and standard Git sources, because we have 'private' Git repos that can't be mirrorred.
Can you simply test your solution by just adding  the  `source 'https://github.com/CocoaPods/Specs'` to your Podfile, and see that this is enough to make the `pod install` fail?

**Comment by mirkoluchi** [comment]
@eyalbe4  Hi, I noticed that in commit  482cda09365e9ff74a0246a5bd71a9f55ec9bf83 you **removed** the following code from **cocoapods_plugin.rb**:

```
Pod::HooksManager.register('cocoapods-art', :source_provider) do |context, options|
    Pod::UI.message 'cocoapods-art received source_provider hook'
    return unless (sources = options['sources'])
    sources.each do |source_name|
        source = create_source_from_name(source_name)
        context.add_source(source)
    end
end
```

It seems to me that this might be the cause of the issue I'm experiencing, since - if the Artifactory sources are not added - `pod install` and `pod update` will not search for specs in them.
I tried to restore that lines of code and re-build and re-install the gem, and it seems working fine.

Was there any reason why you removed that piece of code?

**Comment by eyalbe4** [comment]
@mirkoluchi - thank you for this insight! 
There's a chance that you're right, and the removal of this code introduces this issue with *pod install*.
Can you please try and add the code back, and let us know if this resolves the issue?

**Comment by mirkoluchi** [comment]
@eyalbe4 - Hi, as told in my previous post, `I tried to restore that lines of code and re-build and re-install the gem, and it seems working fine.`.
However, to avoid continuous iteration, could you please test yourself a case that covers the scenario (so using your plugin with Artifactory+Cocoapods master repo+Git private repo)?

Regards,
Mirko

**Comment by eyalbe4** [comment]
Sure @mirkoluchi. I somehow missed the part where you said you tested it and that it worked for you :)
Let me test it as well. Thanks!

**Comment by eyalbe4** [comment]
@mirkoluchi, 
I've added back the code and tested it. Can you please clone and build the code, as an additional verification before we publish it?
I appreciate your ongoing help and my apologies for the delay.

**Comment by Ashton-W** [comment]
@mirkoluchi I think our scenario is the opposite of yours. We use repo-art to source CocoaPods/Specs, and want to use regular sources for our PrivateSpecs repo. Mixing sources like this doesn't work.

We tested this and pushing to a spec with dependencies in artifactory to a git source repo worked.
Basically, this command works now:

```sh
bundle exec pod repo push REDACTEDSpecs REDACTED.podspec --private --sources=art-cocoapods-specs
```

**Comment by mirkoluchi** [comment]
@eyalbe4 
Hi, I tested again and it seems to me it works fine.
Please publish a new release of the gem and - if any new issue will come out - I'll tell you.
Thanks

**Comment by eyalbe4** [comment]
Version 1.0.0 of cocoapods-art with these changes has just been published.
